packages/decoder/lib/decoders.ts

```diff
@@ -133,6 +135,122 @@ export class ProjectDecoder {
     debug("done with allocation");
   }
 
+  /**
+   * **This function is asynchronous.**
+   *
+   * Adds compilations to the decoder after it has started.  Note it is
+   * only presently possible to do this with a `ProjectDecoder` and not
+   * with the other decoder classes.
+   *
+   * @param compilations The compilations to be added.  Take care that these
+   * have IDs distinct from those the decoder already has.
+   */
+  public async addCompilations(
+    compilations: Compilations.Compilation[]
+  ): Promise<void> {
+    this.compilations = [...this.compilations, ...compilations];
+
+    const {
+      definitions: referenceDeclarations,
+      typesByCompilation: userDefinedTypesByCompilation,
+      types: userDefinedTypes
+    } = Compilations.Utils.collectUserDefinedTypesAndTaggedOutputs(
+      compilations
+    );
+
+    Object.assign(this.referenceDeclarations, referenceDeclarations);
+    Object.assign(
+      this.userDefinedTypesByCompilation,
+      userDefinedTypesByCompilation
+    );
+    Object.assign(this.userDefinedTypes, userDefinedTypes);
+
```

Comment:
Is there anything to be done with `this.codeCache` and `this.ensSettings`? Or have they nothing to do with adding new compilations?

Reply:
No, those should be unaffected.